sdk/test-utils/perfstress/src/testProxyHttpClient.ts

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { HttpOperationResponse, URLBuilder } from "@azure/core-http";
+import { DefaultHttpClient, WebResource, WebResourceLike } from "@azure/core-http";
+import {
+  PipelinePolicy,
+  PipelineRequest,
+  PipelineResponse,
+  SendRequest,
+  createDefaultHttpClient,
+  HttpClient,
+  createPipelineRequest
+} from "@azure/core-rest-pipeline";
+
+const paths = {
+  playback: "/playback",
+  record: "/record",
+  start: "/start",
+  stop: "/stop"
+};
+
+export class TestProxyHttpClientV1 extends DefaultHttpClient {
+  private _uri: string;
+  private _httpClient: DefaultHttpClient;
+  private _recordingId?: string;
+  public _mode!: string;
+
+  constructor(uri: string) {
+    super();
+    this._uri = uri;
+    this._httpClient = new DefaultHttpClient();
+  }
+
+  async sendRequest(request: WebResourceLike): Promise<HttpOperationResponse> {
+    if (this._recordingId && (this._mode === "record" || this._mode === "playback")) {
+      request.headers.set("x-recording-id", this._recordingId);
+      request.headers.set("x-recording-mode", this._mode);
+      request.headers.set("x-recording-remove", "false");
+
+      const upstreamUrl = URLBuilder.parse(request.url);
+      const redirectedUrl: URLBuilder = URLBuilder.parse(request.url);
+
+      const providedUrl = URLBuilder.parse(this._uri);
+      redirectedUrl.setHost(providedUrl.getHost());
+      redirectedUrl.setScheme(providedUrl.getScheme());
+      redirectedUrl.setPort(providedUrl.getPort());
+      upstreamUrl.setPath(undefined);
+      request.headers.set("x-recording-upstream-base-uri", upstreamUrl.toString());
+      request.url = redirectedUrl.toString();
+    }
+
+    return await super.sendRequest(request);
+  }
+
+  async startRecording(): Promise<void> {
+    const startUri = this._uri + paths.record + paths.start;
+    const req = this._createRecordingRequest(startUri);
+    const rsp = await this._httpClient.sendRequest(req);
+    if (rsp.status !== 200) {
+      throw new Error("Start request failed.");
+    }
+    const id = rsp.headers.get("x-recording-id");
+    if (!id) {
+      throw new Error("No recording ID returned.");
+    }
+    this._recordingId = id;
+  }
+
+  async stopRecording(): Promise<void> {
+    const stopUri = this._uri + paths.record + paths.stop;
+    const req = this._createRecordingRequest(stopUri);
+    req.headers.set("x-recording-id", this._recordingId!);
+    await this._httpClient.sendRequest(req);
+  }
+
+  async startPlayback(): Promise<void> {
+    const startUri = this._uri + paths.playback + paths.start;
+    const req = this._createRecordingRequest(startUri);
+    req.headers.set("x-recording-id", this._recordingId!);
```

Comment:
I'm a litte confused about the lifecycle of `_recordingId` - I see for the record case it gets set on `startRecording` and it seems like it likewise gets set below for `startPlayback`. What is confusing me is this line, where sometimes we send it when starting playback, but normally I would think it wouldn't be set since playback and record happen in different runs.

Reply:
> wouldn't be set since playback and record happen in different runs.

No, both record and playback happen in the same run. We are not saving the recordings in a file.

Workflow
- test resources are setup
   - hitting the live service
- then start record
   - making a request to the proxy server to start recording
   - proxy server gives a recording id, we'll use this id to save the actual requests and responses
- run the runAsync once
   - proxy-server saves all the requests and responses in memory
- stop record
   - making a request to the proxy server to stop recording
- start playback
   - making a request to the proxy server to start playback
   - we use the same recording-id that we used in the record mode since that's the only way proxy-server knows what requests are supposed to be played back
   - As a response, we get a new recording-id
- run runAsync again 
   - based on the duration, iterations, and parallel options provided for the perf test
   - all the requests in the runAsync method are played back since we have already recorded them before
- when the runAsync loops end, stop playback
   - making a request to the proxy server to stop playing back
- Delete the live resources that we have created before

Reply:
Now I get it! Makes perfect sense for perf testing

Reply:
Thanks for the detailed breakdown of how the Proxy works. It would be great if you include this info in the Proxy section of the README

Reply:
Added in the readme